nifi-extension-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/ListBoxFileMetadataTemplates.java

```diff
@@ -0,0 +1,231 @@
+                        String cleanFieldName = fieldName.startsWith("/") ? fieldName.substring(1) : fieldName;
+                        String fieldValue = metadata.getValue(fieldName).asString();
+                        templateFields.put(cleanFieldName, fieldValue);
+                    }
+                }
+
+                templatesList.add(templateFields);
+            }
+
+            try {
+                try (final OutputStream out = session.write(flowFile);
+                     final BoxMetadataJsonArrayWriter writer = BoxMetadataJsonArrayWriter.create(out)) {
+
+                    // Write each metadata template as a separate JSON object in the array
+                    for (Map<String, Object> templateFields : templatesList) {
+                        writer.write(templateFields);
+                    }
+                }
+
+                final Map<String, String> recordAttributes = new HashMap<>();
+                recordAttributes.put("record.count", String.valueOf(templatesList.size()));
+                recordAttributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
+                recordAttributes.put("box.file.id", fileId);
+                recordAttributes.put("box.metadata.templates.names", String.join(",", templateNames));
+                recordAttributes.put("box.metadata.templates.count", String.valueOf(templatesList.size()));
+                flowFile = session.putAllAttributes(flowFile, recordAttributes);
+
+                session.getProvenanceReporter().receive(flowFile, BoxFileUtils.BOX_URL + fileId);
+                session.transfer(flowFile, REL_SUCCESS);
+            } catch (final IOException e) {
+                getLogger().error("Failed writing metadata templates from file [{}]", fileId, e);
+                flowFile = session.putAttribute(flowFile, ERROR_MESSAGE, e.getMessage());
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+            }
+
+        } catch (final BoxAPIResponseException e) {
+            flowFile = session.putAttribute(flowFile, ERROR_CODE, valueOf(e.getResponseCode()));
+            flowFile = session.putAttribute(flowFile, ERROR_MESSAGE, e.getMessage());
+            if (e.getResponseCode() == 404) {
+                getLogger().warn("Box file with ID {} was not found.", fileId);
+                session.transfer(flowFile, REL_NOT_FOUND);
+            } else {
+                getLogger().error("Couldn't fetch metadata templates from file with id [{}]", fileId, e);
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+            }
+        } catch (final Exception e) {
+            getLogger().error("Failed to process metadata templates for file [{}]", fileId, e);
+            flowFile = session.putAttribute(flowFile, ERROR_MESSAGE, e.getMessage());
+            flowFile = session.penalize(flowFile);
```

Comment:
I don't think I would penalize the flowfile here, if the user wants to retry the flowfile with a penalty duration, the user should use the retry mechanism at framework level instead of using a self loop relationship.